### PR TITLE
Create onyx.update

### PR DESCRIPTION
**Pull request by luacmartins**
### Related Issues

GH_LINK

### CODE REVIEW

**Comment by luacmartins** [comment]
I have read the CLA Document and I hereby sign the CLA

File: lib/Onyx.js [hunk 1]
```diff
@@ -780,6 +780,39 @@ function mergeCollection(collectionKey, collection) {
...
+function update(data) {
+    // First, validate the Onyx object is in the format we expect
+    _.each(data, ({onyxMethod, key, value}) => {
+        if (!_.contains(['set', 'merge'], onyxMethod)) {
+            throw new Error(`Invalid onyxMethod ${onyxMethod} in Onyx update.`);
+        }
+        if (!_.isString(key)) {
+            throw new Error(`Invalid ${typeof key} key provided in Onyx update. Onyx key must be of type string.`);
+        }
+        if (typeof value === 'undefined') {
```
Comment (tgolen) [inline lib/Onyx.js:800]: Use `_.isUndefined()` to be consistent with other underscore methods.

File: lib/Onyx.js [hunk 2]
```diff
@@ -780,6 +780,39 @@ function mergeCollection(collectionKey, collection) {
...
+    });
+
+    _.each(data, ({onyxMethod, key, value}) => {
+        if (onyxMethod === 'set') {
+            set(key, value);
+            return;
+        }
+        if (onyxMethod === 'merge') {
+            merge(key, value);
+        }
```
Comment (tgolen) [inline lib/Onyx.js:812]: NAB: I think doing a `switch()` is slightly better here, and it makes it a bit more extensible for the future.

File: lib/Onyx.js [hunk 3]
```diff
@@ -780,6 +780,39 @@ function mergeCollection(collectionKey, collection) {
...
+            throw new Error(`Invalid ${typeof key} key provided in Onyx update. Onyx key must be of type string.`);
+        }
+        if (typeof value === 'undefined') {
+            throw new Error(`Undefined value passed to key ${key} in Onyx update.`);
+        }
+    });
+
+    _.each(data, ({onyxMethod, key, value}) => {
+        if (onyxMethod === 'set') {
+            set(key, value);
```
Comment (tgolen) [inline lib/Onyx.js:807]: NAB: It might be nice to collect all the things using set and pass them to `multiSet()`, but since we are trying to move away from using `set()` at all, the could be a preoptimization. I kinda wish we had a `multiMerge()` though.

File: tests/unit/onyxTest.js [hunk 4]
```diff
@@ -306,4 +306,80 @@ describe('Onyx', () => {
...
+            Onyx.update(data);
+        } catch (error) {
+            expect(error.message).toEqual('Invalid onyxMethod multiSet in Onyx update.');
+        }
+
+        try {
+            data[1] = {onyxMethod: 'merge', key: true, value: {test2: 'test2'}};
+            Onyx.update(data);
+        } catch (error) {
+            // eslint-disable-next-line max-len
```
Comment (tgolen) [inline tests/unit/onyxTest.js:374]: Is this necessary? It looks like another liner higher up is longer than this, and I thought the length was something like 190, which this line looks to be an OK length.

File: lib/Onyx.js (hunk 1 above)
Comment (tgolen) [reply lib/Onyx.js:800]: Also, I think it's OK for `undefined` to be passed with `set()` is that right? It would be expected to remove the Onyx key entirely.

**Review by tgolen — CHANGES_REQUESTED** [review]
I always comment about this in PRs adding tests, and unfortunately, we don't do this consistently, but I think it GREATLY improves the readability of tests (especially when someone that didn't write them has the task of maintaining them). Could you please add comments in the test that follow this format:

```
// GIVEN <initial state of the system>
...code that sets up the initial state

// WHEN <something happens>
...code that does something

// THEN <explain the final expected state of the system>
...assertions
```

File: lib/Onyx.js (hunk 1 above)
Comment (luacmartins) [reply lib/Onyx.js:800]: That seems right based on [this comment](https://github.com/Expensify/react-native-onyx/blob/8772195796b9a0c9c6fec5a3feeeb214f2b6a5ca/lib/Onyx.js#L631). However, [this line](https://github.com/Expensify/react-native-onyx/blob/8772195796b9a0c9c6fec5a3feeeb214f2b6a5ca/lib/Onyx.js#L634) seems to only check for `null` values. Maybe I'm missing something though.

File: lib/Onyx.js (hunk 2 above)
Comment (luacmartins) [reply lib/Onyx.js:812]: Sounds good. I switch (pun intended)

File: lib/Onyx.js (hunk 3 above)
Comment (luacmartins) [reply lib/Onyx.js:807]: I'm not sure how much of a pre-optimization that is, so I'll skip this for now.

File: tests/unit/onyxTest.js (hunk 4 above)
Comment (tgolen) [reply tests/unit/onyxTest.js:374]: Yea, I agree with not having warnings. I guess I wonder why it's at 120 and not 190. Maybe try updating the eslint config to make sure it's using our latest?

File: lib/Onyx.js (hunk 1 above)
Comment (luacmartins) [reply lib/Onyx.js:800]: I'll remove this check since it seems like we want to support `undefined` values as well.

Comment (tgolen) [reply lib/Onyx.js:800]: Hm, I don't think that's quite right. That line is for merging, and it just means that if a `null` value is passed to `merge()` then it will get ignored (hence, it is "deleted" from the data being merged into Onyx).

The [`setItem()` method in LocalForage](https://localforage.github.io/localForage/#data-api-setitem) doesn't seem to mention what happens if you pass `undefined` as the value, but it might be worth experimenting with to verify the expected behavior!

Comment (luacmartins) [reply lib/Onyx.js]: Hmm that makes sense. I experimented a bit and it seems that passing undefined does remove the key. I'm removing that check then. Thanks!

File: lib/Onyx.js [hunk 5]
```diff
@@ -780,6 +780,39 @@ function mergeCollection(collectionKey, collection) {
         });
 }
 
+/**
+ * Insert API responses and lifecycle data into Onyx
+ *
+ * @param {Object} data An object containing the Onyx data to set
```
Comment (marcochavezf) [inline lib/Onyx.js:786]: This should be an `Array`, no?

```suggestion
 * @param {Array} data An array containing the Onyx data to set
```

File: tests/unit/onyxTest.js (hunk 4 above)
Comment (luacmartins) [reply tests/unit/onyxTest.js:385]: Yea, we are really outdated on eslint-config. Updating it surfaces A LOT of errors, so I'll address that in a separate PR. I'll keep the comment for now then.

File: lib/Onyx.js (hunk 5 above)
Comment (luacmartins) [reply lib/Onyx.js:786]: Good call! Thanks!

File: tests/unit/onyxTest.js [hunk 6]
```diff
@@ -306,4 +306,84 @@ describe('Onyx', () => {
...
+                        onyxMethod: 'merge',
+                        key: ONYX_KEYS.ANOTHER_TEST,
+                        value: {test2: 'test2'},
+                    },
+                ]);
+                return waitForPromisesToResolve();
+            })
+            .then(() => {
+                // THEN the final Onyx state should be {test: 'one', anotherTest: {test1: 'test1', test2: 'test2'}}
+                expect(testKeyValue).toBe('one');
```
[the remark this answers is not in the document]
Comment (luacmartins) [reply tests/unit/onyxTest.js:357]: Haha I was testing something and shouldn't have committed that. I already reverted those changes.

**Comment by luacmartins** [comment]
Updated!

**Review by marcochavezf — APPROVED** [review]
LGTM 
NAB: I think it will be great to add unit tests that set `undefined` and `null` values

**Review by tgolen — CHANGES_REQUESTED** [review]

File: lib/Onyx.js [hunk 7]
```diff
@@ -780,6 +780,36 @@ function mergeCollection(collectionKey, collection) {
...
+/**
+ * Insert API responses and lifecycle data into Onyx
+ *
+ * @param {Array} data An array of objects with shape {onyxMethod: oneOf('set', 'merge'), key: string, value: *}
+ */
+function update(data) {
+    // First, validate the Onyx object is in the format we expect
+    _.each(data, ({onyxMethod, key}) => {
+        if (!_.contains(['set', 'merge'], onyxMethod)) {
+            throw new Error(`Invalid onyxMethod ${onyxMethod} in Onyx update.`);
```
[the remark this answers is not in the document]
Comment (luacmartins) [reply lib/Onyx.js:792]: I'm not sure they are quite the same. If any of these checks fail, we don't run any set/merge at all vs if this happens in the switch, we could run a few set/merges before getting to the bad object. 

I'd imagine that we only want to process the update if the full API response is valid?

[the remark this answers is not in the document]
Comment (luacmartins) [reply lib/Onyx.js:792]: Removing the default case from the switch or just adding a `return` as default triggers lint errors. Options are:

1. Add this error handling, knowing that we should never hit that code.
2. Manually disable the lint error
3. Switch back to if statements
4. Something else?

File: lib/Onyx.js [hunk 8]
```diff
@@ -883,6 +914,7 @@ function applyDecorators() {
     mergeCollection = decorate.decorateWithMetrics(mergeCollection, 'Onyx:mergeCollection');
     getAllKeys = decorate.decorateWithMetrics(getAllKeys, 'Onyx:getAllKeys');
     initializeWithDefaultKeyStates = decorate.decorateWithMetrics(initializeWithDefaultKeyStates, 'Onyx:defaults');
+    update = decorate.decorateWithMetrics(update, 'Onyx:update');
```
Comment (luacmartins) [inline lib/Onyx.js:917]: I had missed adding the update method to the decorators for capturing metrics

File: lib/Onyx.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (luacmartins) [reply lib/Onyx.js:792]: Chatted 1:1 with Tim and we decided to just add a `break` to the default method!

**Comment by luacmartins** [comment]
Updated!

File: tests/unit/onyxTest.js (hunk 4 above)
Comment (luacmartins) [reply tests/unit/onyxTest.js:385]: Got a [PR up](https://github.com/Expensify/react-native-onyx/pull/135) to update eslint-config

File: lib/Onyx.js [hunk 9]
```diff
@@ -780,6 +780,36 @@ function mergeCollection(collectionKey, collection) {
         });
 }
 
+/**
+ * Insert API responses and lifecycle data into Onyx
+ *
+ * @param {Array} data An array of objects with shape {onyxMethod: oneOf('set', 'merge'), key: string, value: *}
```
[the remark this answers is not in the document]
Comment (luacmartins) [reply lib/Onyx.js:786]: That's a good idea. I'm not sure we want to block this PR on adding this reference though. Created an [issue](https://github.com/Expensify/react-native-onyx/issues/138) for it. Thanks!

**Comment by luacmartins** [comment]
Ready for another review

File: lib/Onyx.js [hunk 10]
```diff
@@ -780,6 +780,36 @@ function mergeCollection(collectionKey, collection) {
...
+    _.each(data, ({onyxMethod, key, value}) => {
+        switch (onyxMethod) {
+            case 'set':
+                set(key, value);
+                break;
+            case 'merge':
+                merge(key, value);
+                break;
+            default:
+                break;
```
Comment (Julesssss) [inline lib/Onyx.js:808]: Usually, I prefer switch/case, but this seems more verbose than two simple conditional checks. NAB though

**Review by Julesssss — APPROVED** [review]
Looks good to me, left a small comment but it's not a blocker. Feel free to resolve nobody else agrees 🙂

File: lib/Onyx.js (hunk 10 above)
Comment (luacmartins) [reply lib/Onyx.js:808]: Haha I initially had it as a series of conditionals, but changed it because of @tgolen's comments [here](https://github.com/Expensify/react-native-onyx/pull/133#discussion_r879765933). I don't have strong feeling either way.

Comment (luacmartins) [reply lib/Onyx.js:808]: Sounds good! Let's not switch this then 🤣

File: lib/Onyx.js [hunk 11]
```diff
@@ -780,6 +780,36 @@ function mergeCollection(collectionKey, collection) {
...
+ * @param {Array} data An array of objects with shape {onyxMethod: oneOf('set', 'merge'), key: string, value: *}
+ */
+function update(data) {
+    // First, validate the Onyx object is in the format we expect
+    _.each(data, ({onyxMethod, key}) => {
+        if (!_.contains(['set', 'merge'], onyxMethod)) {
+            throw new Error(`Invalid onyxMethod ${onyxMethod} in Onyx update.`);
+        }
+        if (!_.isString(key)) {
+            throw new Error(`Invalid ${typeof key} key provided in Onyx update. Onyx key must be of type string.`);
```
[the remark this answers is not in the document]
Comment (luacmartins) [reply lib/Onyx.js:795]: That's a good idea. We could certainly add that as a follow up if we come across this issue!

File: lib/Onyx.js [hunk 12]
```diff
@@ -780,6 +780,36 @@ function mergeCollection(collectionKey, collection) {
         });
 }
 
+/**
+ * Insert API responses and lifecycle data into Onyx
+ *
+ * @param {Array} data An array of objects with shape {onyxMethod: oneOf('set', 'merge'), key: string, value: *}
+ */
+function update(data) {
```
[the remark this answers is not in the document]
Comment (luacmartins) [reply lib/Onyx.js:788]: @parasharrajat thanks for investigating this. So can we just return the respective methods [here](https://github.com/Expensify/react-native-onyx/blob/2eb3d74e7b581007d7c4dcc332b5bbc027145ce2/lib/Onyx.js#L837-L852)? e.g.

```
case 'merge':
    return merge(key, value);
```
